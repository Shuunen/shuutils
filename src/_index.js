'use strict'

import { getTimestamp, getTimestampMs } from './timestamps'

module.exports = { getTimestamp, getTimestampMs }