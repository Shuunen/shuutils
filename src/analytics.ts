/* eslint-disable @typescript-eslint/prefer-readonly-parameter-types */
import { sleep } from './functions'
import { Logger } from './logger'

type AnalyticsInstance = {
  identify: (userId: string, additionalData?: Record<string, unknown>) => Promise<void>
  page: () => Promise<void>
  track: (event: string, additionalData?: Record<string, unknown>) => Promise<void>
}

/**
 * Generic analytics wrapper, can be used with any analytics third party library
 * @example const analytics = new Analytics()
 * @example analytics.setupInMemory('my-app')
 * @example analytics.track('invoice-download')
 * @example analytics.track('click', { category: '...', label: '...', ... })
 */
// eslint-disable-next-line no-restricted-syntax
export class Analytics {
  private $app = ''
  private $instance: AnalyticsInstance | undefined = undefined
  private readonly $logger = new Logger()
  public readonly pile: string[] = []

  /**
   * Identify a user
   * @param userId - The user, like '123456', 'super-admin', etc.
   * @param additionalData - Additional data to identify the user, like `{ email: '...', age: 35, ... }`
   */
  public identify(userId: string, additionalData?: Record<string, unknown>) {
    if (this.$instance === undefined) throw new Error('analytics : identify failed, analytics not setup')
    this.$logger.info(`analytics : identify user "${userId}" on app "${this.$app}"`)
    void this.$instance.identify(userId, additionalData)
  }

  /**
   * Track a page view
   */
  public page() {
    if (this.$instance === undefined) throw new Error('analytics : track page failed, analytics not setup')
    this.$logger.debug(`analytics : track page for app "${this.$app}"`)
    void this.$instance.page()
  }

  /**
   * Setup analytics with in memory tracking, useful for tests
   * @param app - The name of the app, like 'mlm-react-front', 'cosmos-front', etc.
   */
  public setupInMemory(app: string) {
    this.$app = app
    // eslint-disable-next-line jsdoc/require-jsdoc
    const addToPile = async (item: string, data?: Record<string, unknown>) => {
      this.$logger.debug(`analytics : add "${item}" to pile`, data)
      await sleep(1)
      this.pile.push(item)
      this.$logger.debug(`analytics : pile for app "${app}"`, this.pile)
    }
    this.$instance = {
      // eslint-disable-next-line jsdoc/require-jsdoc
      identify: async (userId: string, additionalData?: Record<string, unknown>) => {
        await addToPile(`identify:${userId}`, additionalData)
      },
      // eslint-disable-next-line jsdoc/require-jsdoc
      page: async () => {
        await addToPile('page')
      },
      // eslint-disable-next-line jsdoc/require-jsdoc
      track: async (event: string, additionalData?: Record<string, unknown>) => {
        await addToPile(`track:${event}`, additionalData)
      },
    }
    this.$logger.info(`analytics : setup app "${app}" with in memory tracking`)
  }

  /**
   * Track an event
   * @param event - The event name like 'click', 'download-invoice', etc.
   * @param additionalData - Additional data to track the event, like `{ category: '...', label: '...', ... }`
   */
  public track(event: string, additionalData?: Record<string, unknown>) {
    if (this.$instance === undefined) throw new Error('analytics : track event failed, analytics not setup')
    this.$logger.debug(`analytics : track event "${event}" for app "${this.$app}"`)
    void this.$instance.track(event, additionalData)
  }
}
