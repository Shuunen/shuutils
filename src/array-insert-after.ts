import { nbAfter, nbNoIndex } from './constants'
import { clone } from './object-clone'

/**
 * Add a value to an array if it doesn't already exist
 * @param array the array to add the value to
 * @param item the value to add
 * @param value the value to add
 * @returns the array with the value added (if it didn't already exist)
 */
export function insertValueAfterItem<T> (array: T[], item: T, value: T) {
  const index = array.indexOf(item)
  if (index === nbNoIndex) return array
  const arrayCopy = clone(array)
  const start = Math.min(index + nbAfter, arrayCopy.length)
  arrayCopy.splice(start, 0, value)
  return arrayCopy
}
