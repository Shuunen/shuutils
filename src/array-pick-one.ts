/**
 * Return an item from an array
 * @param items like : ["great", "place", "pine"]
 * @returns item like : "pine"
 */
export function pickOne<Type>(items: readonly Type[]) {
  if (items.length === 0) throw new Error('Array is empty')
  // eslint-disable-next-line @typescript-eslint/consistent-type-assertions
  return items[Math.floor(Math.random() * items.length)] as Type
}
