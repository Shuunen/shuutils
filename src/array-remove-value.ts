import { clone } from './object-clone'

/**
 * Remove a value from an array
 * @param array the array to remove the value from
 * @param value the value to remove
 * @returns a copy of the array without the value
 */
export function removeValueFromArray<Type>(array: readonly Type[], value: Type) {
  if (!array.includes(value)) return array
  const arrayCopy = clone(array)
  arrayCopy.splice(arrayCopy.indexOf(value), 1)
  return arrayCopy
}
