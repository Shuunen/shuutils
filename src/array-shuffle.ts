import { Nb } from './constants'
import { clone } from './object-clone'

/**
 * Randomize an array using the Fisher-Yates algorithm
 * source : https://dev.to/codebubb/how-to-shuffle-an-array-in-javascript-2ikj
 * @param input the array to shuffle
 * @returns a shuffled clone of the input array
 */
export function shuffleArray<T> (input: T[]) {
  const array = clone(input)
  // eslint-disable-next-line for-direction
  for (let indexA = array.length - 1; indexA > 0; indexA += Number(Nb.Previous)) {
    const indexB = Math.floor(Math.random() * (indexA + 1))
    const temporary = array[indexA]
    array[indexA] = array[indexB] as T // eslint-disable-line @typescript-eslint/consistent-type-assertions
    array[indexB] = temporary as T // eslint-disable-line @typescript-eslint/consistent-type-assertions
  }
  return array
}
