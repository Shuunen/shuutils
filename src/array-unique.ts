/**
 * Return an array with unique values
 * @param items the array to check/clean/reduce
 * @returns the array with unique values
 */
export function arrayUnique<Type>(items: readonly Type[]) {
  return Array.from(new Set(items))
}
