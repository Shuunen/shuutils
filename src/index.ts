export * from './array-insert-after'
export * from './array-pick-one'
export * from './array-remove-value'
export * from './array-shuffle'
export * from './array-unique'
export * from './browser-clipboard'
export * from './browser-fill-like-human'
export * from './browser-inject-styles'
export * from './browser-page-change'
export * from './browser-scout'
export * from './browser-toast'
export * from './colors'
export * from './constants'
export * from './date-relative'
export * from './dates'
export * from './debounce'
export * from './dom'
export * from './environment'
export * from './events'
export * from './expect-type'
export * from './functions'
export * from './logger'
export * from './memoize'
export * from './number-random'
export * from './number-ranged-score'
export * from './number-round'
export * from './object-cap-nesting'
export * from './object-clone'
export * from './object-equal'
export * from './object-flatten'
export * from './object-serializer'
export * from './object-sort'
export * from './objects'
export * from './state'
export * from './storage'
export * from './string-remove-accents'
export * from './strings'
export * from './testing-align-for-snap'
export * from './throttle'
export * from './timestamps'
export * from './types'
