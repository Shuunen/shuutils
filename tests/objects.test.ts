import { strictEqual as equal } from 'assert'
import { test } from 'uvu'
import { access, byProperty, clone, copy, flatten, genClass } from '../src'
import { check } from './utils'

const person = { name: 'John', age: 21, details: { favoriteFood: 'sushi' } }
const persons = ['John', 'Fanny']

test('copy object', function () {
  const personCopy = copy(person)
  personCopy.age = 42
  equal(person.age, 21)
  equal(personCopy.age, 42)
})

test('copy-less object', function () {
  const personCopy = person
  personCopy.age = 42
  equal(person.age, 42)
  equal(personCopy.age, 42)
})

test('copy array', function () {
  const personsCopy = copy(persons)
  personsCopy[1] = 'Bob'
  equal(persons[1], 'Fanny')
  equal(personsCopy[1], 'Bob')
})

test('copy-less array', function () {
  const personsCopy = persons
  personsCopy[1] = 'Bob'
  equal(persons[1], 'Bob')
  equal(personsCopy[1], 'Bob')
})

check('clone is a better name', copy(person), clone(person))

check('nested existing property access', access(person, 'details.favoriteFood'), 'sushi')
check('nested non-existing property access', typeof access(person, 'details.favoriteDrink'), 'undefined')
check('nested non-existing property access with a default value', access(person, 'details.favoriteDrink', 'wine'), 'wine')
check('access a non-nested property', access({ name: 'John Cena' }, 'name'), 'John Cena')
check('access a non-existing non-nested property', typeof access({ name: 'John Cena' }, 'age'), 'undefined')
check('access a non-nested property after an undefined property', access({ age: undefined, name: 'John Cena' }, 'name'), 'John Cena')

check('flatten an object', flatten(person), { 'age': 21, 'details.favoriteFood': 'sushi', 'name': 'John' })
check('flatten an object with a custom root path', flatten(person, 'person'), { 'person.age': 21, 'person.details.favoriteFood': 'sushi', 'person.name': 'John' })
check('flatten an object containing an array', flatten({ name: 'John', collection: ['pikachu', 'drake'] }), { 'name': 'John', 'collection[0]': 'pikachu', 'collection[1]': 'drake' })

const users = [{ name: 'John', age: 21, pic: 'wow.png' }, { name: 'Albert', age: 42 }, { name: 'Sam', age: 22 }, { name: 'Birgit', age: 11 }]
check('sort objects by property without order does not sort', users.sort(byProperty('name'))[0].name, 'John')
check('sort objects by property with asc order does sort', users.sort(byProperty('name', 'asc'))[0].name, 'Albert')
check('sort objects by property with desc order does sort', users.sort(byProperty('name', 'desc'))[0].name, 'Sam')
check('sort objects by property even if some does not have it', users.sort(byProperty('pic', 'asc'))[0].pic, 'wow.png')

const object3 = { 'superFun': true, 'notFun': false, 'pretty-good': true, 'size': 'large' }
check('class generator empty', genClass({}), '')
check('class generator object', genClass(object3), 'superFun pretty-good size-large')
check('class generator object & specific keys', genClass(object3, ['pretty-good']), 'pretty-good')
check('class generator object & specific keys & custom class', genClass(object3, ['pretty-good'], ['nice ok']), 'nice ok pretty-good')

test.run()
