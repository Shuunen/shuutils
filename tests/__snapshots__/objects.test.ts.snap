// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`objectSort C sort a complex object with null, undefined, etc 1`] = `
{
  "alpaca": {
    "c1": 11,
    "c2": 22,
    "c3": 33,
    "c4": undefined,
  },
  "keyA": undefined,
  "keyB": 2,
  "keyC": 3,
  "names": [
    "john",
    null,
    [
      {
        "b2": 2,
        "c3": 3,
      },
    ],
    "eddy",
  ],
  "zebra": {
    "z1": 11,
    "z2": null,
    "z3": 33,
  },
}
`;
